Mathlib/Combinatorics/SetFamily/Shadow.lean

```diff
@@ -93,24 +96,79 @@ theorem erase_mem_shadow (hs : s ∈ 𝒜) (ha : a ∈ s) : erase s a ∈ ∂ 
   mem_shadow_iff.2 ⟨s, hs, a, ha, rfl⟩
 #align finset.erase_mem_shadow Finset.erase_mem_shadow
 
+/-- `t ∈ ∂𝒜` iff `t` is exactly one element less than something from `𝒜`.
+
+See also `Finset.mem_shadow_iff_exists_mem_card_add_one`. -/
+lemma mem_shadow_iff_exists_sdiff : t ∈ ∂ 𝒜 ↔ ∃ s ∈ 𝒜, t ⊆ s ∧ (s \ t).card = 1 := by
+  simp_rw [mem_shadow_iff, ←covby_iff_card_sdiff_eq_one, covby_iff_exists_erase, eq_comm]
+
 /-- `t` is in the shadow of `𝒜` iff we can add an element to it so that the resulting finset is in
 `𝒜`. -/
-theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ (a : _) (_ : a ∉ s), insert a s ∈ 𝒜 := by
-  refine' mem_shadow_iff.trans ⟨_, _⟩
-  · rintro ⟨s, hs, a, ha, rfl⟩
-    refine' ⟨a, not_mem_erase a s, _⟩
-    rwa [insert_erase ha]
-  · rintro ⟨a, ha, hs⟩
-    exact ⟨insert a s, hs, a, mem_insert_self _ _, erase_insert ha⟩
+theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ a, a ∉ s ∧ insert a s ∈ 𝒜 := by
+  simp_rw [mem_shadow_iff_exists_sdiff, ←covby_iff_card_sdiff_eq_one, covby_iff_exists_insert]
+  aesop
 #align finset.mem_shadow_iff_insert_mem Finset.mem_shadow_iff_insert_mem
 
+/-- `s ∈ ∂ 𝒜` iff `s` is exactly one element less than something from `𝒜`.
+
+See also `Finset.mem_shadow_iff_exists_sdiff`. -/
+lemma mem_shadow_iff_exists_mem_card_add_one :
+    s ∈ ∂ 𝒜 ↔ ∃ t ∈ 𝒜, s ⊆ t ∧ t.card = s.card + 1 := by
+  refine mem_shadow_iff_exists_sdiff.trans $ exists_congr fun t ↦ and_congr_right fun _ ↦
+    and_congr_right fun hst ↦ ?_
+  rw [card_sdiff hst, tsub_eq_iff_eq_add_of_le, add_comm]
+  exact card_mono hst
+#align finset.mem_shadow_iff_exists_mem_card_add_one Finset.mem_shadow_iff_exists_mem_card_add_one
+
+/-- `t ∈ ∂^k 𝒜` iff `t` is exactly `k` elements less than something from `𝒜`.
+
+See also `Finset.mem_shadow_iff_exists_mem_card_add`. -/
+lemma mem_shadow_iterate_iff_exists_sdiff :
+    s ∈ ∂^[k] 𝒜 ↔ ∃ t ∈ 𝒜, s ⊆ t ∧ (t \ s).card = k := by
+  induction' k with k ih generalizing 𝒜 s
+  · simp only [sdiff_eq_empty_iff_subset, Function.iterate_zero, id.def, card_eq_zero, exists_prop]
+    refine' ⟨fun p ↦ ⟨s, p, Subset.rfl, Subset.rfl⟩, _⟩
+    rintro ⟨t, hs, hst, hts⟩
+    rwa [subset_antisymm hst hts]
+  simp only [exists_prop, Function.comp_apply, Function.iterate_succ, ih,
+    mem_shadow_iff_insert_mem]
+  clear ih
+  constructor
+  · rintro ⟨t, ⟨a, ha, hs⟩, hts, rfl⟩
+    refine' ⟨_, hs, hts.trans $ subset_insert _ _, _⟩
+    rw [insert_sdiff_of_not_mem _ $ not_mem_mono hts ha,
+      card_insert_of_not_mem $ not_mem_mono (sdiff_subset _ _) ha]
+  · rintro ⟨t, hs, hts, hk⟩
+    obtain ⟨a, ha⟩ : (t \ s).Nonempty := by rw [←card_pos, hk]; exact Nat.succ_pos _
+    refine' ⟨erase t a, ⟨a, not_mem_erase _ _, _⟩, subset_erase.2 ⟨hts, (mem_sdiff.1 ha).2⟩, _⟩
+    · rwa [insert_erase (mem_sdiff.1 ha).1]
+    · rw [erase_sdiff_comm, card_erase_of_mem ha, hk, succ_sub_one]
```

Comment:
```suggestion

lemma mem_shadow_iterate_iff_exists_third :
    s ∈ ∂^[k] 𝒜 ↔ ∃ u : Finset α, u.card = k ∧ Disjoint s u ∧ s ∪ u ∈ 𝒜 := by
  induction' k with k ih generalizing s
  · simp
  simp only [mem_shadow_iff_insert_mem, ih, Function.iterate_succ_apply', card_eq_succ]
  aesop

/-- `t ∈ ∂^k 𝒜` iff `t` is exactly `k` elements less than something from `𝒜`.

See also `Finset.mem_shadow_iff_exists_mem_card_add`. -/
lemma mem_shadow_iterate_iff_exists_sdiff :
    s ∈ ∂^[k] 𝒜 ↔ ∃ t ∈ 𝒜, s ⊆ t ∧ (t \ s).card = k := by
  rw [mem_shadow_iterate_iff_exists_third]
  constructor
  · rintro ⟨u, rfl, hsu, hsuA⟩
    exact ⟨_, hsuA, subset_union_left s u, by rw [union_sdiff_cancel_left hsu]⟩
  rintro ⟨t, htA, hst, rfl⟩
  refine ⟨t \ s, rfl, disjoint_sdiff, ?_⟩
  rwa [union_sdiff_self_eq_union, union_eq_right.2 hst]
```
Here's a way to refactor which makes the induction part a bit clearer, gives you another useful-looking statement generalising `mem_shadow_iff_insert_mem`, and makes the proof of `mem_shadow_iterate_iff_exists_sdiff` simpler. Feel free to change the name though